### ISSUE TITLE
Horovod on multinode multicore cpu cluster

### ISSUE THREAD

**Issue by UditGupta10** [horovod/horovod#626]
The description of the cluster (cray xc40) is as:
Each node has 2 CPU sockets with 12 cores each, 128GB RAM and connected using Cray Aries interconnect.

For the program:
from __future__ import print_function
import horovod.tensorflow as hvd
hvd.init()
print('local_rank=%d, rank=%d, size=%d' % (hvd.local_rank(), hvd.rank(), hvd.size()))

I spawn 4 processes on 2 nodes  (2 processes per node) using
aprun -j 1 -n 4 -N 2 python test.py
and get following output:
local_rank=0, rank=2, size=4
local_rank=1, rank=1, size=4
local_rank=0, rank=0, size=4
local_rank=1, rank=3, size=4

So, this means that each core is assigned a distinct rank. 
For the case of multi gpu I understand that each gpu is allotted a rank and holds the parameters which are averaged after computation of a batch.
In the case of multinode multicore cpu cluster if each core is assigned a rank then where will the parameters be stored since a core does not have its own memory and how does parameter averaging happen?

**Comment by alsrgv** [comment]
@UditGupta10, based on your description, I'm assuming this cluster does not have GPU - is that correct?

In this configuration, each CPU socket will run a training process.  CPU typically has plenty of RAM it's attached to.  A copy of the parameters will be owned by each process, and they'll exchange gradients with each other using Horovod.

**Comment by UditGupta10** [comment]
Yes, the cluster doesn't have GPUs.
I am a little confused, each node has 2 sockets and each socket has 12 cores. You mentioned that "each CPU socket will run a training process" then this "A copy of the parameters will be owned by each process". From the code that I ran (mentioned in earlier comment), each process has a different rank. Is the optimisation on multiple cores done implicitly? Please clarify the the multicore operation.

**Comment by alsrgv** [comment]
@UditGupta10, without more context, I can only say that you run two processes per node and that each process will own a copy of its parameters.  I should clarify that whether each process will only be executed by each physical CPU or not depends on MPI implementation & configuration that you're using.  Can you share more details about how exactly you submit the job, which MPI are you using and all that?

**Comment by UditGupta10** [comment]
I am using cray cluster which uses aprun instead of mpirun.
In my case 1 node has 2 CPU sockets, each socket has 12 cores. As per my understanding Tensorflow optimally utilizes the multiple cores present in a node.
I want to run horovod on multiple nodes such that each node uses maximum resources (cores) available.
Please suggest the right value to use for flags.
Following flags are available:
-n: Number of processing elements PEs required for the application (pes)
-N: Number of PEs to place per node
-d: Number of CPU cores required for each PE and its threads
-j: Number of CPUs to use per compute unit (bulldozer core-module)
-cc: Bind PEs to CPU cores.

**Comment by alihydri** [comment]
@alsrgv , can you help me with the details of data distribution method? I have one host and multiple processing  nodes (all CPU) with shared storage units, but only in host-node fashion (not node-node fashion). Can I put the images on different storage units ( ex: half on shared storage between host and node1 and the other half on shared storage between the host and node2)? If so, will Horovod use local data for processing or just randomly uses data from both storage units?

**Comment by abditag2** [comment]
Horovod does not deal with data distribution and it is entirely up to your scripts how to feed the data into each worker.

To create a mental model of how to distribute the data, think about it in this way. You want to partition your entire data to equal chunks for each horovod worker. Now, with shared storage, you can have each worker to read their own chunk from the original data source. If you dont have shared storage, you can copy each chunk to the local storage of the worker. Or you can have the entire data on the local storage and read only the correct partition based the rank of the worker.

**Comment by alihydri** [comment]
Thank you @abditag2. 
Does that mean in case of local storage, the nodes  stop training after processing all of their input data chunks? I had the notion that nodes exchange input data as well as weights and parameters. 
And what's the proper way to allocate parts of shared directory or data to a specific node?

**Comment by abditag2** [comment]
For local storage, you can set the number of epochs -- which should be
identical on all the workers. So you can train for as long as you like and
then yes, the nodes will stop.

For shared storage, or if you want to read only your chunk for your worker
from a large datasets, we recommend using petastorm. Petastorm takes care
of chunking the data and uses multiple threads for reading data which
increases performance.

**Comment by abditag2** [comment]
You can look at this example for using petastorm with tf: https://github.com/horovod/horovod/blob/9f87459ead9ebb7331e1cd9cf8e9a5543ecfb784/examples/keras_spark_rossmann.py#L453

There are pytorch examples on the petastorm github repo as well.
